qiskit/quantum_info/__init__.py

```diff
@@ -163,11 +155,4 @@
     state_fidelity,
     negativity,
 )
-from .synthesis import (
```

Comment:
Sorry, I missed this before: since these objects were importable from `qiskit.quantum_info` before, I believe they should remain importable throughout the deprecation period. But what's annoying is that any unrelated import from `quantum_info` will also trigger this deprecation warning... 

I'm not sure if we can nicely solve this, without introducing intermediary classes like
```python
from qiskit.synthesis import OneQubitEulerDecomposer as NewOneQubitEulerDecomposer

class OneQubitEulerDecomposer(NewOneQubitEulerDecomposer):
    def __init__(...):
        # warn here
```
Maybe @jakelishman has an idea?

Reply:
I think that filtering the warning from this specific import in the quantum info init would work in this case.

Reply:
You can use a module-level `__getattr__` to lazily effect the deprecation, so the warning gets raised only if somebody attempts to import the bad name specifically.  For example, `qiskit.circuit.__init__` used to do this:
```python
_DEPRECATED_NAMES = {
    "Int1": "qiskit.circuit.classicalfunction.types",
    "Int2": "qiskit.circuit.classicalfunction.types",
    "classical_function": "qiskit.circuit.classicalfunction",
    "BooleanExpression": "qiskit.circuit.classicalfunction",
}


def __getattr__(name):
    if name in _DEPRECATED_NAMES:
        import importlib
        import warnings

        module_name = _DEPRECATED_NAMES[name]
        warnings.warn(
            f"Accessing '{name}' from '{__name__}' is deprecated since Qiskit Terra 0.22 "
            f"and will be removed in 0.23.  Import from '{module_name}' instead. "
            "This will require installing 'tweedledum' as an optional dependency from Terra 0.23.",
            DeprecationWarning,
            stacklevel=2,
        )
        return getattr(importlib.import_module(module_name), name)
    raise AttributeError(f"module '{__name__}' has no attribute '{name}'")
```

That meant that `import qiskit.circuit` raised no warnings, but `from qiskit.circuit import Int1` or `qiskit.circuit.Int1` would warn, and successfully find the object in the new location.

Reply:
Apparently that message was from a time before the strengthened deprecation policy, although we didn't remember to remove it until pretty recently haha.

Reply:
I am a bit confused. Could you please explain what should I do with the functions moved from quantum_info to the synthesis library?

Reply:
The `qiskit/quantum_info/synthesis/__init__.py` file stays like it is (though is `PendingDeprecationWarning` correct? - I'm not sure what the plan is, and I'm not up-to-date on this PR sorry).  The stuff you've moved already also stays like it is.

At the bottom of `qiskit/quantum_info/__init__.py` we'd add:

```python
_DEPRECATED_NAMES = {
    "OneQubitEulerDecomposer": "qiskit.synthesis",
    "TwoQubitBasisDecomposer": "qiskit.synthesis",
    "XXDecomposer": "qiskit.synthesis",
    "decompose_clifford": "qiskit.synthesis",
    "two_qubit_cnot_decompose": "qiskit.synthesis",
}


def __getattr__(name):
    if name in _DEPRECATED_NAMES:
        import importlib
        import warnings

        module_name = _DEPRECATED_NAMES[name]
        warnings.warn(
            f"Accessing '{name}' from '{__name__}' is deprecated since Qiskit <whatever>"
            f" and will be removed in <whatever>.  Import from '{module_name}' instead.",
            DeprecationWarning,
            stacklevel=2,
        )
        return getattr(importlib.import_module(module_name), name)
    raise AttributeError(f"module '{__name__}' has no attribute '{name}'")
```

That means that `import qiskit.quantum_info` will not attempt to import `qiskit.quantum_info.synthesis`, nor to import `qiskit.synthesis`, but if a user explicitly says `from qiskit.quantum_info import TwoQubitBasisDecomposer`, then they'll get a warning that says

> Accessing 'TwoQubitBasisDecomposer' from 'qiskit.quantum_info' is deprecated since Qiskit <whatever> and will be removed in <whatever>.  Import from 'qiskit.synthesis' instead."

and then the `TwoQubitBasisDecomposer` object will be correctly imported from `qiskit.synthesis` automatically like it suggests.